src/ja/server/scheduler/default_policies.py

```diff
@@ -29,51 +45,129 @@ class DefaultJobDistributionPolicyBase(JobDistributionPolicy, ABC):
                        available_machines: List[WorkMachine]) -> Optional[Tuple[WorkMachine, List[Job]]]:
-        # Should iterate over all available machines and select the best one depending on _assign_machine_cost
-        pass
+        chosen_machine: Tuple[WorkMachine, List[Job]] = None
+        chosen_cost: float = 0
+
+        for machine in available_machines:
+            job_entries = \
+                list(filter(lambda e: ((e.assigned_machine.uid if e.assigned_machine else None) == machine.uid),
+                     distribution))
+
+            result = self._assign_machine_cost(job, machine, job_entries)
+            if result:
+                (cost, preempted) = result
+                if not chosen_machine or cost < chosen_cost:
+                    chosen_cost = cost
+                    chosen_machine = (machine, preempted)
+        return chosen_machine
 
 
 class DefaultNonPreemptiveDistributionPolicy(DefaultJobDistributionPolicyBase):
+    _cpu_threads_w = 2.0
+    _memory_w = 1.0
     """
     Default distribution policy for non-preemptive scheduling.
     """
+    def __init__(self, cost_function: CostFunction):
+        """
+        Initialize the distribution policy.
+        @param cost_function The cost function to use to determine urgent jobs.
+        """
+        self._cost_func = cost_function
+
     def _assign_machine_cost(self,
-                             job: Job,
+                             job: DatabaseJobEntry,
                              machine: WorkMachine,
-                             existing_jobs: List[Job]) -> Optional[Tuple[int, List[Job]]]:
-        pass
+                             existing_jobs: List[DatabaseJobEntry]) -> Optional[Tuple[float, List[Job]]]:
+        if not self._check_machine_feasible(job.job, machine, free_only=True):
+            return None
+        if machine.resources.free_resources.swap < job.job.docker_constraints.memory:
+            if self._cost_func.calculate_cost(job) > self._cost_func.preempting_threshold:
+                return None
+
+        after_allocation = machine.resources.free_resources - self._get_job_allocation(job.job)
+        return (after_allocation.cpu_threads * self._cpu_threads_w + after_allocation.memory * self._memory_w, [])
 
 
 class DefaultBlockingDistributionPolicy(DefaultJobDistributionPolicyBase):
     """
     Default distribution policy for blocking jobs.
     """
     def _assign_machine_cost(self,
-                             job: Job,
+                             job: DatabaseJobEntry,
                              machine: WorkMachine,
-                             existing_jobs: List[Job]) -> Optional[Tuple[int, List[Job]]]:
-        pass
+                             existing_jobs: List[DatabaseJobEntry]) -> Optional[Tuple[float, List[Job]]]:
+        if not self._check_machine_feasible(job.job, machine, free_only=False):
+            return None
+        return (len(existing_jobs), [])
```

Comment:
But then the question is, how do I pick which jobs need to finish? Start with largest by CPU? Largest by RAM?

The idea about total number of jobs is just because the more jobs there are, they are likely small to be on the same machine, so we'll need to wait for more of then to finish.

Reply:
Hmm, and by the way, if you look at the algorithm itself, you'll notice I allow a blocking job to be assigned on a machine different from the machine blocked for it (in which case the blocked machine should of course be marked as "free")

Reply:
>But then the question is, how do I pick which jobs need to finish?

What would work in any case is to just try out combinations of jobs and see if them finishing would give you the resources you need.
So you would try if any 1 job finishing gives you enough resources, then if any 2 jobs finishing gives you enough resources, and so on.
If you're concerned about the runtime of this approach we could also do a heuristic that only checks if any 1 or 2 jobs finishing would be enough to give you the required resources.

>The idea about total number of jobs is just because the more jobs there are, they are likely small to be on the same machine, so we'll need to wait for more of them to finish.

This assumes that all jobs on a machine need to finish in order to run the job.
However, if only a small percentage of jobs needs to finish in order to run the blocking job it would actually be better to pick a machine with a large number of jobs because it would be more likely that any of them will finish in the near future.

Reply:
I changed the logic so that now it tries to use the expected number of jobs that need to finish so that we can schedule our job, please take a look again and see if it's good enough.